### PR TITLE
Replace Readline, bitshares-core issue 673

### PR DESCRIPTION
**Pull request by jmjatlanta**
This is for the issue https://github.com/bitshares/bitshares-core/issues/673

This is the Pull Request for the bitshares-fc portion. The submodule https://github.com/troglobit/editline was added to this library.

The related pull request that handles the bitshares-core portion is at https://github.com/bitshares/bitshares-core/pull/685

### CODE REVIEW

File: src/rpc/cli.cpp [hunk 1]
```diff
@@ -134,53 +120,73 @@ char * dupstr (const char* s) {
...
-
-   auto& cmd = cli_commands();
-
-   while( list_index < cmd.size() ) 
-   {
-      name = cmd[list_index].c_str();
-      list_index++;
-
-      if (strncmp (name, text, len) == 0)
-         return (dupstr(name));
```
Comment (pmconrad) [inline src/rpc/cli.cpp:155]: With this, the dupstr() function will no longer be used an should be removed as well.

File: src/rpc/cli.cpp [hunk 2]
```diff
@@ -134,53 +120,73 @@ char * dupstr (const char* s) {
...
-   }
-
-   /* If no names matched, then return NULL. */
-   return ((char *)NULL);
+    int matchlen = 0;
+    int count = 0;
+    const char* method_name;
+
+    auto& cmd = cli_commands();
+    for (auto it = cmd.begin(); it != cmd.end(); it++) {
```
Comment (pmconrad) [inline src/rpc/cli.cpp]: Using a foreach loop over cmd would make the code look better. Same in next function.

File: src/rpc/cli.cpp [hunk 3]
```diff
@@ -134,53 +120,73 @@ char * dupstr (const char* s) {
...
+    	int partlen = strlen (token); /* Part of token */
```
Comment (pmconrad) [inline src/rpc/cli.cpp]: Move this out of the loop. token is never modified, no need to re-count in every cycle.

File: src/rpc/cli.cpp [hunk 4]
```diff
@@ -201,7 +209,7 @@ void cli::getline( const fc::string& prompt, fc::string& line)
          line_read = readline(prompt.c_str());
          if( line_read == nullptr )
             FC_THROW_EXCEPTION( fc::eof_exception, "" );
-         rl_bind_key( '\t', rl_complete );
+         //el_bind_key( '\t', rl_complete );
```
Comment (pmconrad) [inline src/rpc/cli.cpp]: Why is this commented out? Is \t the default?

File: src/rpc/cli.cpp (hunk 1 above)
Comment (jmjatlanta) [reply src/rpc/cli.cpp:155]: Removed

File: src/rpc/cli.cpp (hunk 2 above)
Comment (jmjatlanta) [reply src/rpc/cli.cpp]: Switched to foreach in both places

File: src/rpc/cli.cpp (hunk 3 above)
Comment (jmjatlanta) [reply src/rpc/cli.cpp]: Moved partlen initialization to outside the loop

File: src/rpc/cli.cpp (hunk 4 above)
Comment (jmjatlanta) [reply src/rpc/cli.cpp]: Yes, \t is the default. Removed the line.

File: .gitmodules [hunk 5]
```diff
@@ -7,3 +7,6 @@
 [submodule "vendor/websocketpp"]
     path = vendor/websocketpp
     url = https://github.com/zaphoyd/websocketpp.git
+[submodule "vendor/editline"]
+	path = vendor/editline
```
Comment (abitmore) [inline .gitmodules]: Please replace the tabs with spaces.

File: src/rpc/cli.cpp (hunk 4 above)
Comment (abitmore) [reply src/rpc/cli.cpp]: Please push your new commit :)

File: .gitmodules (hunk 5 above)
Comment (jmjatlanta) [reply .gitmodules]: Replaced tab with 4 spaces

**Comment by jmjatlanta** [comment]
I lost a commit somehow. Changes have been redone. 

Note: The new editline library has a liberal license, but still requires documentation changes. See the license at https://github.com/troglobit/editline/blob/master/LICENSE

File: src/rpc/cli.cpp (hunk 4 above)
Comment (jmjatlanta) [reply src/rpc/cli.cpp]: Commit lost somewhere. Redone.

**Review by pmconrad — APPROVED** [review]

**Comment by pmconrad** [comment]
Thanks!

**Pull request merged**